### PR TITLE
Add tag interning deviate

### PR DESCRIPTION
**Pull request by robert-milan**
This updates defByTagSet to use a new Md5Hash struct in idx. This should reduce memory consumption for keys in the map by the size of the string data returned by NameWithTags().

Md5Hash is 16 bytes (2x uint64), which is the same size as the overhead used for storing a string. However, now we aren't actually storing the string data.

This PR also changes most of the index to use pointers for MetricDefinitions and Archives that exist inside the index. Their lifecycles are locked with one another so there is no reason to use values. Most of the underlying data structures that are manipulated are slices and regardless of whether or not a pointer is used they are all still operating on the same backing arrays.

### CODE REVIEW

File: idx/memory/memory.go [hunk 1]
```diff
@@ -137,35 +138,39 @@ func (t *TagIndex) delTagId(name, value uintptr, id schema.MKey, m *MemoryIdx) {
...
+	hashedName := def.NameWithTagsHash()
+	if _, ok = orgDefs[hashedName]; !ok {
+		orgDefs[hashedName] = make(map[*idx.MetricDefinition]struct{}, 1)
 	}
-	orgDefs[fullName][def] = struct{}{}
+
+	if _, ok = orgDefs[hashedName][def]; !ok {
+		orgDefs[hashedName][def] = struct{}{}
+	}
+	// orgDefs[fullName][def] = struct{}{}
```
Comment (replay) [inline idx/memory/memory.go]: commented code

File: docker/docker-dev-custom-cfg-kafka/metrictank.ini [hunk 2]
```diff
@@ -385,7 +385,7 @@ disable-initial-host-lookup = false
 [memory-idx]
 enabled = false
 # enables/disables querying based on tags
-tag-support = false
+tag-support = true
```
Comment (replay) [inline docker/docker-dev-custom-cfg-kafka/metrictank.ini]: was that committed intentionally? or did you only change that for testing?

Comment (robert-milan) [reply docker/docker-dev-custom-cfg-kafka/metrictank.ini]: That was for testing, forgot to revert it before commit

**Comment by Dieterbe** [comment]
please in OP describe what this is about....

File: idx/metric_definition.go [hunk 3]
```diff
@@ -354,9 +355,13 @@ func (md *MetricDefinition) NameWithTags() string {
...
-func (md *MetricDefinition) NameWithTagsHash() string {
-	buffer := bytes.NewBufferString(md.NameWithTags())
-	return fmt.Sprintf("%x", md5.Sum(buffer.Bytes()))
+func (md *MetricDefinition) NameWithTagsHash() Md5Hash {
+	md5Sum := md5.Sum(bytes.NewBufferString(md.NameWithTags()).Bytes())
+	ret := Md5Hash{
+		Upper: binary.LittleEndian.Uint64(md5Sum[:8]),
+		Lower: binary.LittleEndian.Uint64(md5Sum[8:]),
+	}
+	return ret
```
Comment (replay) [inline idx/metric_definition.go]: I'd add some function `idx.Md5HashFromString(string) Md5Hash`, since this snippet of specific and repetitive

Comment (robert-milan) [reply idx/metric_definition.go]: Not sure I follow how it is repetitive. I feel like it is more closely tied to the MetricDefinition than anything else. It only appears in one more place which is in memory.go.

File: idx/idx_gen.go [hunk 4]
```diff
@@ -209,7 +243,141 @@ func (z *Archive) UnmarshalMsg(bts []byte) (o []byte, err error) {
...
+				err = msgp.WrapError(err)
+				return
+			}
+		}
+	}
+	return
+}
+
+// EncodeMsg implements msgp.Encodable
+func (z Md5Hash) EncodeMsg(en *msgp.Writer) (err error) {
```
Comment (Dieterbe) [inline idx/idx_gen.go]: why do we need msgp support for this? it's only used in 1 in-memory-only datastructure, right?

File: idx/metric_definition.go (hunk 3 above)
Comment (Dieterbe) [reply idx/metric_definition.go]: i can see both sides of the argument. let's just leave as is and move on

File: idx/idx_gen.go (hunk 4 above)
Comment (robert-milan) [reply idx/idx_gen.go]: We definitely do not need msgp support for this. I will add an exclusion.

**Pull request merged**